external-crates/move/crates/move-vm-runtime/src/execution/interpreter/state.rs

```diff
@@ -434,47 +434,82 @@ impl Stack {
 
 impl CallStack {
     /// Create a new empty call stack.
-    fn new() -> Self {
-        CallStack(vec![])
-    }
+    pub fn new(
+        resolver: ModuleDefinitionResolver,
+        function: ArenaPointer<Function>,
+        ty_args: Vec<Type>,
+        args: Vec<Value>,
+    ) -> PartialVMResult<Self> {
+        let mut heap = MachineHeap::new();
 
-    /// Push a `Frame` on the call stack.
-    fn push(&mut self, frame: CallFrame) -> ::std::result::Result<(), CallFrame> {
-        if self.0.len() < CALL_STACK_SIZE_LIMIT {
-            self.0.push(frame);
-            Ok(())
-        } else {
-            Err(frame)
-        }
-    }
+        let fun_ref = function.to_ref();
+        let stack_frame = heap.allocate_stack_frame(args, fun_ref.local_count())?;
+        let current_frame = CallFrame {
+            pc: 0,
+            stack_frame,
+            resolver,
+            function,
+            ty_args,
+        };
 
-    /// Pop a `Frame` off the call stack.
-    fn pop(&mut self) -> Option<CallFrame> {
-        self.0.pop()
+        Ok(Self {
+            current_frame,
+            heap,
+            frames: vec![],
+        })
     }
-}
 
-impl CallFrame {
     /// Create a new `Frame` given a `Function` and the function's `ty_args` and `args`.
-    /// This loads the locals, padding appropriately.
+    /// This loads the locals, padding appropriately, and sets the call stack's current frame.
     #[inline]
-    pub fn new(
+    pub fn push_call(
```

Comment:
Nit on naming: thoughts on calling this `push_frame` for parity with `pop_frame`? That or possibly renaming `pop_frame` to something like `pop_call` or something -- no strong opinion on names so long as they're consistent.